lib/sycamore/sycamore/query/execution/operations.py

```diff
@@ -31,6 +33,112 @@
 NUM_TEXT_CHARS_GENERATE = 2500
 
 
+class QuestionAnsweringSummarizer:
+    def __init__(self, llm: LLM, question: str):
+        self.llm = llm
+        self.question = question
+
+    def __call__(self, text: str) -> str:
+        messages = SummarizeDataMessagesPrompt(question=self.question, text=text).as_messages()
+        prompt_kwargs = {"messages": messages}
+
+        t0 = time.time()
+        # call to LLM
+        summary = self.llm.generate(prompt_kwargs=prompt_kwargs, llm_kwargs={"temperature": 0})
+        t1 = time.time()
+        log.info(f"Summarizer took {t1 - t0} seconds to generate summary.")
+
+        return summary
+
+
+def collapse(text: str, tokens_per_chunk: int, tokenizer: Tokenizer, summarizer_fn: Callable[[str], str]) -> str:
+    """
+    Collapses text iteratively, summarizing the first chunk and incorporating it in the summary for the next chunk.
+
+    Args:
+        text: Text to collapse.
+        chunk_size: Size of each chunk.
+        tokenizer: Tokenizer to use for counting against max_tokens.
+
+    Returns:
+        List of chunks.
+    """
+    tokens = tokenizer.tokenize(text)
+    total = len(tokens)
+    if total <= tokens_per_chunk:
+        return text
+    done = False
+    i = 0
+    additional = i + tokens_per_chunk
+    cur_summary = ""
+    while not done:
+        input = ""
+        if cur_summary:
+            input = f"{cur_summary}\n"
+        input += "".join(tokens[i : i + additional])
+        print(f"input size: {len(input)}")
+        cur_summary = summarizer_fn(input)
+        assert (
+            len(cur_summary) <= tokens_per_chunk
+        ), f"Summarizer output is longer than input chunk {len(cur_summary)} > {tokens_per_chunk} !!!"
+        print(f"summary to chunk ratio: {len(cur_summary) / tokens_per_chunk}")
+        i += additional
+        remaining = tokens_per_chunk - len(cur_summary)
+        additional = min(remaining, total - i)
+        if additional == 0:
+            break
+
+    return cur_summary
+
+
+class DocumentSummarizer(Summarizer):
```

Comment:
would it make sense for this to live in docset?

Reply:
I think so.  I will move it there.